subcommand/server-acl-init/command_ent_test.go

```diff
@@ -0,0 +1,436 @@
+// +build enterprise
+
+package serveraclinit
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/mitchellh/cli"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"testing"
+)
+
+// Test that ACL policies get updated if namespaces config changes.
+func TestRun_ACLPolicyUpdates(t *testing.T) {
+	t.Parallel()
+	k8s, testAgent := completeEnterpriseSetup(t, resourcePrefix)
+	defer testAgent.Stop()
+	require := require.New(t)
+
+	ui := cli.NewMockUi()
+	firstRunArgs := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-create-client-token",
+		"-allow-dns",
+		"-create-mesh-gateway-token",
+		"-create-sync-token",
+		"-create-inject-namespace-token",
+		"-expected-replicas=1",
+		"-allow-dns",
+	}
+	// Our second run, we're going to update from namespaces disabled to
+	// namespaces enabled with a single destination ns.
+	secondRunArgs := append(firstRunArgs,
+		"-enable-namespaces",
+		"-consul-sync-destination-namespace=sync",
+		"-consul-inject-destination-namespace=dest")
+
+	// Run the command first to populate the policies.
+	cmd := Command{
+		UI:        ui,
+		clientset: k8s,
+	}
+	responseCode := cmd.Run(firstRunArgs)
+	require.Equal(0, responseCode, ui.ErrorWriter.String())
+
+	bootToken := getBootToken(t, k8s, resourcePrefix)
+	consul, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+
+	// Check that the expected policies were created.
+	expectedPolicies := []string{
+		"dns-policy",
+		"client-token",
+		"catalog-sync-token",
+		"connect-inject-token",
+		"mesh-gateway-token",
+	}
+	policies, _, err := consul.ACL().PolicyList(nil)
+	require.NoError(err)
+	for _, expected := range expectedPolicies {
+		found := false
+		for _, p := range policies {
```

Comment:
What about doing a single loop through `policies` first to put them in a map of policy name -> policy struct? Then when looping through `expectedPolicies`, you can look them up directly.

Reply:
Good call, switched to:

```go
	// Collect the actual polices into a map to make it easier to assert
	// on their existence and contents.
	actualPolicies := make(map[string]string)
	for _, p := range policies {
		policy, _, err := consul.ACL().PolicyRead(p.ID, nil)
		require.NoError(err)
		actualPolicies[p.Name] = policy.Rules
	}
	for _, expected := range expectedPolicies {
		actPolicy, ok := actualPolicies[expected]
		require.True(ok, "Did not find policy %s", expected)
		// We assert that the policy doesn't have any namespace config
		// in it because later that's what we're using to test that it
		// got updated.
		require.NotContains(actPolicy, "namespace")
	}
```